ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -852,8 +857,12 @@ protected void maybeProcessCallbacks() {
               routerMetrics.getFirstDataChunkLatencyMs.update(System.currentTimeMillis() - submissionTimeMs);
             }
             routerMetrics.getDataChunkLatencyMs.update(System.currentTimeMillis() - initializedTimeMs);
-            chunkIndexToBuf.put(chunkIndex, filterChunkToRange(result.getDecryptedBlobContent()));
-            numChunksRetrieved.incrementAndGet();
+            ByteBuf decryptedContent = result.getDecryptedBlobContent();
```

Comment:
The decryption runs in a separate thread pool asynchronously, so that progress with other operations of the single-threaded `OperationController` is not delayed. Should we create a similar scheduler for decryption as well?

Reply:
Encryption uses another thread.  It does not make the operation faster faster or higher throughput but the code is much more complicated.  Let me explain.

It's important to have 2 layers of threads.  The IO threads and the business logic threads.  The IO threads is managed by Netty.  So when Netty invokes a callback, it's calling using IO thread.  Ambry handler needs to respond to Netty callback ASAP so there is no IO blocking.  That means the callback handler (running in Netty IO thread) should pickup the object and pass it to Ambry business logic threads, and return right way to stop taking more IO thread time.  The business logic threads should do what the processing.  If business logic needs to do disk IO, it should be asynchronous.  If it needs to loop through the content, it should be synchronous.  Yes, synchronous!  It doesn't help if Thread A calls Thread B and wait for Thread B to complete.  Encryption and compression are effectively looping through the content.  It does not help for one business logic thread to pass that work to another business logic thread (in the same thread pool).

Our operation manager already created the business logic thread pool to handle Netty requests and responses, so when put operation executes, they are already executing it the business logic thread.  There is no need to create more threads.  Our router code can be much simpler.  Let's work on cleaning it up.